File: C#/Code/SPTemplate.Operations/ListOperations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.SharePoint;
using System.Collections;
using System.IO;
using Microsoft.SharePoint.Utilities;
using System.Xml;
using System.Web;
using System.Net;
using Microsoft.SharePoint.Administration;

namespace $safeprojectname$
{
    internal class ListOperations : BaseOperation
    {
        public ListOperations(SPWeb objSPWeb) : base(objSPWeb) { }

        public Guid CreateList(out Boolean IsExisting, String strListName, String strDescription, ArrayList xmlFieldCollection, SPListTemplateType listType, Boolean requireCheckout)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            SPList objList;
            Guid objListGUID = default(Guid);
            IsExisting = false;
            try
            {
                if (!IsListExisting(strListName))
                {
                    //create list here
                    objListGUID = objSPWeb.Lists.Add(strListName, strDescription, listType);
                    objList = objSPWeb.Lists.GetList(objListGUID, false);
                    if (requireCheckout)
                    {
                        objSPWeb.AllowUnsafeUpdates = true;
                        objList.EnableVersioning = true;
                        objList.EnableMinorVersions = true;
                        objList.ForceCheckout = true;
                        objList.EnableFolderCreation = false;
                        objList.Update();
                    }
                }
                else
                {
                    IsExisting = true;
                    objList = objSPWeb.Lists.TryGetList(strListName);
                    objListGUID = objList.ID;
                }
                foreach (String strfieldXML in xmlFieldCollection)
                {
                    try
                    {
                        objList.Fields.AddFieldAsXml(strfieldXML, true, SPAddFieldOptions.AddFieldCheckDisplayName);
                    }
                    catch { }
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
            return objListGUID;
        }

        public Guid CreateList(out Boolean IsExisting, String strListName, String strDescription, ArrayList xmlFieldCollection, SPListTemplateType listType)
        {
            return CreateList(out IsExisting, strListName,  strDescription, xmlFieldCollection, listType, false);
        }

        public SPField CreateSiteColumn(out Boolean IsExisting, String xmlSiteColXml)
        {
            SPField objSPField = null;
            SPWeb objSPRootWeb = this.CurrentWeb.Site.RootWeb;
            IsExisting = false;
            XmlDocument objFieldXml = null;
            String strFieldStaticName = String.Empty;
            String strColTitle = String.Empty;
            try
            {
                objFieldXml = new XmlDocument();
                objFieldXml.LoadXml(xmlSiteColXml);
                if (objFieldXml != null && objFieldXml.HasChildNodes)
                {
                    strFieldStaticName = objFieldXml.ChildNodes[0].Attributes["StaticName"].Value;
                    if (!String.IsNullOrEmpty(strFieldStaticName))
                    {
                        objSPField = objSPRootWeb.Fields.TryGetFieldByStaticName(strFieldStaticName);
                        if (objSPField == null)
                        {
                            strFieldStaticName = objSPRootWeb.Fields.AddFieldAsXml(xmlSiteColXml, true, SPAddFieldOptions.AddFieldCheckDisplayName);
                            if (!String.IsNullOrEmpty(strFieldStaticName))
                            {
                                objSPField = objSPRootWeb.Fields.GetFieldByInternalName(strFieldStaticName);
                            }
                        }
                        else
                        {
                            IsExisting = true;
                        }
                    }
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
            return objSPField;
        }

        public Boolean IsListExisting(String listName)
        {
            Boolean listExists = false;
            SPList objSPList = null;
            SPWeb objSPWeb;
            objSPWeb = this.CurrentWeb;

            objSPList = objSPWeb.Lists.TryGetList(listName);
            if (objSPList != null)
            {
                listExists = true;
            }
            return listExists;
        }

        
        public void GetListItem(String strListName, String strQueryXml, out SPListItem objSPListItem, SPWeb objSPWeb)
        {
            SPListItemCollection objSPListItemCollection = null;

            try
            {
                objSPListItem = null;
                objSPListItemCollection = GetListItems(strListName, strQueryXml, objSPWeb);
                if (objSPListItemCollection != null && objSPListItemCollection.Count > 0)
                {
                    objSPListItem = objSPListItemCollection[0];
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
        }

        public SPListItemCollection GetListItems(String strListName, String strQueryXml)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            return GetListItems(strListName, strQueryXml, objSPWeb);
        }

        public SPListItemCollection GetListItems(String strListName, String strQueryXml, SPWeb objSPWeb)
        {
            SPList objSPList = null;
            SPQuery objSPQuery;
            SPListItemCollection objSPListItemCollection = null;

            try
            {

                objSPList = objSPWeb.Lists.TryGetList(strListName);

                if (objSPList != null)
                {
                    if (!String.IsNullOrEmpty(strQueryXml))
                    {
                        objSPQuery = new SPQuery();
                        objSPQuery.ViewAttributes = "Scope='Recursive'";
                        objSPQuery.Query = strQueryXml;
                        objSPListItemCollection = objSPList.GetItems(objSPQuery);
                    }
                }
            }
            catch (Exception Ex)
            {

            }
            return objSPListItemCollection;
        }

        public SPListItem GetListItemByID(String strListName, Int32 ID)
        {
            SPWeb objSPWeb = this.CurrentWeb.Site.RootWeb;
            return GetListItemByID(strListName, ID, objSPWeb);
        }

        public SPListItem GetListItemByListID(String ListID, Int32 ItemID)
        {
            String strListName = String.Empty;
            SPWeb objSPWeb = this.CurrentWeb;
            strListName = GetListNameByListID(ListID);
            return GetListItemByID(strListName, ItemID, objSPWeb);
        }

        public SPListItem GetListItemByListID(String ListID, Int32 ItemID, SPWeb objSPWeb)
        {
            String strListName = String.Empty;
            strListName = GetListNameByListID(ListID);
            return GetListItemByID(strListName, ItemID, objSPWeb);
        }

      
        public String GetListNameByListID(String ListID)
        {
            String strListName = String.Empty;
            SPWeb objSPWeb = this.CurrentWeb;
            SPList objSPList = null;
            try
            {
                objSPList = objSPWeb.Lists.GetList(new Guid(ListID), false, false);
                if (objSPList != null)
                {
                    strListName = objSPList.Title;
                }
            }
            catch (Exception Ex)
            {
                return String.Empty;
            }
            return strListName;
        }

        public SPListItem GetListItemByID(String strListName, Int32 ID, SPWeb objSPWeb)
        {
            SPList objSPList = null;
            SPQuery objSPQuery;
            SPListItemCollection objSPListItemCollection = null;
            SPListItem objSPListItem = null;
            String strQueryXml = String.Empty;
            try
            {
                objSPList = objSPWeb.Lists.TryGetList(strListName);
                if (objSPList != null)
                {
                    strQueryXml = @"<Where><Eq><FieldRef Name='ID' /><Value Type='Counter'>" + ID + "</Value></Eq></Where>";
                    objSPQuery = new SPQuery();
                    objSPQuery.ViewAttributes = "Scope='Recursive'";
                    objSPQuery.Query = strQueryXml;
                    objSPListItemCollection = objSPList.GetItems(objSPQuery);
                    if (objSPListItemCollection.Count > 0)
                    {
                        objSPListItem = objSPListItemCollection[0];
                    }
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
            return objSPListItem;
        }

        public void AddItem(string strListName, out SPListItem objSPListItem)
        {
            SPList objSPList = null;
            SPWeb objSPWeb = null;
            objSPListItem = null;
            try
            {
                objSPWeb = this.CurrentWeb;
                objSPList = objSPWeb.Lists.TryGetList(strListName);
                if (objSPList != null)
                {
                    objSPListItem = objSPList.Items.Add();
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
        }

        public SPField GetSPFieldByName(String strListName, String strfieldName)
        {
            SPField objSPField = null;
            SPList objSPList = null;
            try
            {
                objSPList = this.CurrentWeb.Lists[strListName];
                if (objSPList != null)
                {
                    objSPField = objSPList.Fields.GetField(strfieldName);
                }
            }
            catch (Exception Ex)
            {
                return null;
            }
            return objSPField;
        }

        public SPField GetSPFieldByName(SPFieldCollection fieldColl, String strfieldName)
        {
            SPField objSPField = null;

            try
            {
                if (fieldColl != null)
                {
                    objSPField = fieldColl.GetField(strfieldName);
                }
            }
            catch (Exception Ex)
            {
                return null;
            }

            return objSPField;
        }

        public void DeleteList(String strListName)
        {
            SPList objSPList = null;
            SPWeb objSPWeb = null;
            try
            {
                objSPWeb = this.CurrentWeb;
                if (IsListExisting(strListName))
                {
                    objSPWeb.AllowUnsafeUpdates = true;
                    objSPList = objSPWeb.Lists.TryGetList(strListName);
                    objSPList.Delete();
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
        }

        public Boolean DeleteItemByID(String strListName, Int32 ItemId)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            return DeleteItemByID(strListName, ItemId, objSPWeb);
        }

        public Boolean DeleteItemByTitle(String strListName, String strTitle)
        {
            String strqueryXml = String.Empty;
            Boolean IsSucess = false;
            SPListItem objSPListItem = null;
            try
            {
                strqueryXml = @"<Where><Eq><FieldRef Name='Title' /><Value Type='Text'>" + strTitle + "</Value></Eq></Where>";
                GetListItem(strListName, strqueryXml, out objSPListItem, this.CurrentWeb);
                if (objSPListItem != null)
                {
                    objSPListItem.Delete();
                }
            }
            catch (Exception Ex)
            {
            }
            return IsSucess;
        }

        public Boolean DeleteItemByID(String strListName, Int32 ItemId, SPWeb objSPWeb)
        {
            Boolean isSucess = false;
            SPListItem objSPListItem = null;
            try
            {
                objSPListItem = GetListItemByID(strListName, ItemId);
                if (objSPListItem != null)
                {
                    objSPListItem.Recycle();
                    isSucess = true;
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
            return isSucess;
        }

     
        public bool IsPageExists(String strPath)
        {
            Boolean IsExists = false;
            SPWeb objSPWeb;
            try
            {
                objSPWeb = this.CurrentWeb;
                if (objSPWeb.GetFile(strPath).Exists)
                {
                    IsExists = true;
                }
            }
            catch (Exception Ex)
            {

            }
            return IsExists;
        }

        public bool IsListItemExists(string strListName, string strFileNameWithoutextension)
        {
            SPWeb web = this.CurrentWeb.Site.RootWeb;
            SPList objSPList = null;
            SPListItemCollection objSPListItemCollection = null;
            Boolean boolReturn = false;
            try
            {
                objSPList = web.Lists[strListName];
                objSPListItemCollection = objSPList.Items;
                foreach (SPListItem item in objSPListItemCollection)
                {
                    if (item.DisplayName.ToLower().Equals(strFileNameWithoutextension.ToLower()))
                    {
                        boolReturn = true;
                        break;
                    }
                }
            }
            catch (Exception Ex)
            {
                throw Ex;
            }
            return boolReturn;
        }

        public int GetNewListItemID(string strListName, string strFileNameWithoutextension)
        {
            SPWeb web = this.CurrentWeb.Site.RootWeb;
            SPList objSPList = null;
            SPListItemCollection objSPListItemCollection = null;
            int intListItemID = 0;
            try
            {
                objSPList = web.Lists[strListName];
                objSPListItemCollection = objSPList.Items;
                foreach (SPListItem item in objSPListItemCollection)
                {
                    if (item.DisplayName.Equals(strFileNameWithoutextension))
                    {
                        intListItemID = item.ID;
                    }
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return intListItemID;
        }

        public String GetFileNameById(Int32 ItemID, String strListName)
        {
            SPWeb objSPRootWeb = this.CurrentWeb.Site.RootWeb;
            SPListItem objSPListItem = null;
            String strFileName = String.Empty;
            try
            {
                objSPListItem = GetListItemByID(strListName, ItemID, objSPRootWeb);
                if (objSPListItem != null)
                {
                    strFileName = objSPListItem.File.Name;
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return strFileName;
        }

        public Boolean DeleteItemPermanentlyByID(String strListName, Int32 ItemId)
        {
            Boolean isSucess = false;
            SPListItem objSPListItem = null;
            try
            {
                objSPListItem = GetListItemByID(strListName, ItemId);
                if (objSPListItem != null)
                {
                    objSPListItem.Delete();
                    isSucess = true;
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }
            return isSucess;
        }

        public SPList GetListByListID(String ListID)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            SPList objSPList = null;
            try
            {
                objSPList = objSPWeb.Lists.GetList(new Guid(ListID), true);

            }
            catch (Exception Ex)
            {
                return null;
            }
            return objSPList;
        }

        public SPFieldCollection GetListFieldsByListName(String listName)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            SPFieldCollection fieldColl = null;
            try
            {
                SPList list = objSPWeb.Lists.TryGetList(listName);
                if (list != null)
                {
                    fieldColl = list.Fields;
                }
            }
            catch (Exception Ex)
            {
                return null;
            }
            return fieldColl;
        }

        public SPFieldCollection GetContentTypeFieldsByContentTypeID(String contentTypeID)
        {
            SPWeb objSPWeb = this.CurrentWeb;
            SPFieldCollection fieldColl = null;

            try
            {
                SPContentTypeId cTypeID = new SPContentTypeId(contentTypeID);

                SPContentType contentType = objSPWeb.AvailableContentTypes[cTypeID];

                if (contentType != null)
                {
                    fieldColl = contentType.Fields;
                }
            }
            catch (Exception Ex)
            {
                return null;
            }
            return fieldColl;
        }


      
        public SPList GetListByName(String listName)
        {
            if (!IsListExisting(listName)) return null;

            SPWeb objSPWeb;
            objSPWeb = this.CurrentWeb;

            return objSPWeb.Lists.TryGetList(listName);
        }

        public String GetListIDByListName(String listName)
        {
            String listID = string.Empty;
            SPList objSPList = null;
            SPWeb objSPWeb;
            objSPWeb = this.CurrentWeb;

            objSPList = objSPWeb.Lists.TryGetList(listName);
            if (objSPList != null)
            {
                listID = objSPList.ID.ToString();
            }
            return listID;
        }

        public String GetListURL(String ListName, bool serverRelative = false)
        {
            SPList list = GetListByName(ListName);
            if (null == list) return String.Empty;

            var listUrl = list.DefaultViewUrl;
            if (list is SPDocumentLibrary)
            {
                var idx = listUrl.IndexOf("Forms");
                listUrl = listUrl.Remove(idx);
                if (listUrl.EndsWith("/"))
                    listUrl = listUrl.Remove(listUrl.LastIndexOf("/"));
            }
            else
            {
                var idx = listUrl.LastIndexOf("/");
                listUrl = listUrl.Remove(idx);
            }
            if (!serverRelative)
            {
                var serverUrl = list.ParentWeb.Url;
                serverUrl = serverUrl.Substring(0, serverUrl.IndexOf("/", 7));
                listUrl = serverUrl + listUrl;
            }

            return listUrl;

        }

   
      
        public void PerformMinorCheckin(String strListName, Int32 ItemID, String strMessage)
        {
            SPListItem objListItem = null;
            ListOperations objListOperations = new ListOperations(this.CurrentWeb);
            try
            {
                objListItem = objListOperations.GetListItemByID(strListName, ItemID);
                if (objListItem.ParentList.EnableVersioning)
                {
                    if (objListItem.File.CheckOutType == SPFile.SPCheckOutType.Online)
                    {
                        objListItem.File.CheckIn(strMessage, SPCheckinType.MinorCheckIn);
                    }
                }
            }
            catch { }
        }
    }
}

File: trunk/VB/Code/SPTemplate.Operations/ListOperations.vb
Imports Microsoft.SharePoint
Imports System.Xml

Public Class ListOperations
    Inherits BaseOperation
    Public Sub New(ByVal objSPWeb As SPWeb)
        MyBase.New(objSPWeb)
    End Sub

    Public Sub AddItem(ByVal strListName As String, ByRef objSPListItem As SPListItem)
        Dim objSPList As SPList = Nothing
        Dim objSPWeb As SPWeb = Nothing
        objSPListItem = Nothing
        Try
            objSPWeb = Me.CurrentWeb
            objSPList = objSPWeb.Lists.TryGetList(strListName)
            If objSPList IsNot Nothing Then
                objSPListItem = objSPList.Items.Add()
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
    End Sub

    Public Function CreateList(ByRef IsExisting As [Boolean], ByVal strListName As [String], ByVal strDescription As [String], ByVal xmlFieldCollection As ArrayList, ByVal listType As SPListTemplateType, ByVal requireCheckout As [Boolean]) As Guid
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Dim objList As SPList
        Dim objListGUID As Guid = Nothing
        IsExisting = False
        Try
            If Not IsListExisting(strListName) Then
                'create list here
                objListGUID = objSPWeb.Lists.Add(strListName, strDescription, listType)
                objList = objSPWeb.Lists.GetList(objListGUID, False)
                If requireCheckout Then
                    objSPWeb.AllowUnsafeUpdates = True
                    objList.EnableVersioning = True
                    objList.EnableMinorVersions = True
                    objList.ForceCheckout = True
                    objList.EnableFolderCreation = False
                    objList.Update()
                End If
            Else
                IsExisting = True
                objList = objSPWeb.Lists.TryGetList(strListName)
                objListGUID = objList.ID
            End If
            For Each strfieldXML As [String] In xmlFieldCollection
                Try
                    objList.Fields.AddFieldAsXml(strfieldXML, True, SPAddFieldOptions.AddFieldCheckDisplayName)
                Catch
                End Try
            Next
        Catch Ex As Exception
            Throw Ex
        End Try
        Return objListGUID
    End Function

    Public Function CreateList(ByRef IsExisting As [Boolean], ByVal strListName As [String], ByVal strDescription As [String], ByVal xmlFieldCollection As ArrayList, ByVal listType As SPListTemplateType) As Guid
        Return CreateList(IsExisting, strListName, strDescription, xmlFieldCollection, listType, False)
    End Function

    Public Function CreateSiteColumn(ByRef IsExisting As [Boolean], ByVal xmlSiteColXml As [String]) As SPField
        Dim objSPField As SPField = Nothing
        Dim objSPRootWeb As SPWeb = Me.CurrentWeb.Site.RootWeb
        IsExisting = False
        Dim objFieldXml As XmlDocument = Nothing
        Dim strFieldStaticName As [String] = [String].Empty
        Dim strColTitle As [String] = [String].Empty
        Try
            objFieldXml = New XmlDocument()
            objFieldXml.LoadXml(xmlSiteColXml)
            If objFieldXml IsNot Nothing AndAlso objFieldXml.HasChildNodes Then
                strFieldStaticName = objFieldXml.ChildNodes(0).Attributes("StaticName").Value
                If Not [String].IsNullOrEmpty(strFieldStaticName) Then
                    objSPField = objSPRootWeb.Fields.TryGetFieldByStaticName(strFieldStaticName)
                    If objSPField Is Nothing Then
                        strFieldStaticName = objSPRootWeb.Fields.AddFieldAsXml(xmlSiteColXml, True, SPAddFieldOptions.AddFieldCheckDisplayName)
                        If Not [String].IsNullOrEmpty(strFieldStaticName) Then
                            objSPField = objSPRootWeb.Fields.GetFieldByInternalName(strFieldStaticName)
                        End If
                    Else
                        IsExisting = True
                    End If
                End If
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
        Return objSPField
    End Function

    Public Function DeleteItemByID(ByVal strListName As [String], ByVal ItemId As Int32, ByVal objSPWeb As SPWeb) As [Boolean]
        Dim isSucess As [Boolean] = False
        Dim objSPListItem As SPListItem = Nothing
        Try
            objSPListItem = GetListItemByID(strListName, ItemId)
            If objSPListItem IsNot Nothing Then
                objSPListItem.Recycle()
                isSucess = True
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
        Return isSucess
    End Function

    Public Function DeleteItemByID(ByVal strListName As [String], ByVal ItemId As Int32) As [Boolean]
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Return DeleteItemByID(strListName, ItemId, objSPWeb)
    End Function

    Public Function DeleteItemByTitle(ByVal strListName As [String], ByVal strTitle As [String]) As [Boolean]
        Dim strqueryXml As [String] = [String].Empty
        Dim IsSucess As [Boolean] = False
        Dim objSPListItem As SPListItem = Nothing
        Try
            strqueryXml = "<Where><Eq><FieldRef Name='Title' /><Value Type='Text'>" & Convert.ToString(strTitle) & "</Value></Eq></Where>"
            GetListItem(strListName, strqueryXml, objSPListItem, Me.CurrentWeb)
            If objSPListItem IsNot Nothing Then
                objSPListItem.Delete()
            End If
        Catch Ex As Exception
        End Try
        Return IsSucess
    End Function

    Public Function DeleteItemPermanentlyByID(ByVal strListName As [String], ByVal ItemId As Int32) As [Boolean]
        Dim isSucess As [Boolean] = False
        Dim objSPListItem As SPListItem = Nothing
        Try
            objSPListItem = GetListItemByID(strListName, ItemId)
            If objSPListItem IsNot Nothing Then
                objSPListItem.Delete()
                isSucess = True
            End If
        Catch ex As Exception
            Throw ex
        End Try
        Return isSucess
    End Function

    Public Sub DeleteList(ByVal strListName As [String])
        Dim objSPList As SPList = Nothing
        Dim objSPWeb As SPWeb = Nothing
        Try
            objSPWeb = Me.CurrentWeb
            If IsListExisting(strListName) Then
                objSPWeb.AllowUnsafeUpdates = True
                objSPList = objSPWeb.Lists.TryGetList(strListName)
                objSPList.Delete()
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
    End Sub

    Public Function GetContentTypeFieldsByContentTypeID(ByVal contentTypeID As [String]) As SPFieldCollection
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Dim fieldColl As SPFieldCollection = Nothing

        Try
            Dim cTypeID As New SPContentTypeId(contentTypeID)

            Dim contentType As SPContentType = objSPWeb.AvailableContentTypes(cTypeID)

            If contentType IsNot Nothing Then
                fieldColl = contentType.Fields
            End If
        Catch Ex As Exception
            Return Nothing
        End Try
        Return fieldColl
    End Function

    Public Function GetFileNameById(ByVal ItemID As Int32, ByVal strListName As [String]) As [String]
        Dim objSPRootWeb As SPWeb = Me.CurrentWeb.Site.RootWeb
        Dim objSPListItem As SPListItem = Nothing
        Dim strFileName As [String] = [String].Empty
        Try
            objSPListItem = GetListItemByID(strListName, ItemID, objSPRootWeb)
            If objSPListItem IsNot Nothing Then
                strFileName = objSPListItem.File.Name
            End If
        Catch ex As Exception
            Throw ex
        End Try
        Return strFileName
    End Function

    Public Function GetListByListID(ByVal ListID As [String]) As SPList
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Dim objSPList As SPList = Nothing
        Try

            objSPList = objSPWeb.Lists.GetList(New Guid(ListID), True)
        Catch Ex As Exception
            Return Nothing
        End Try
        Return objSPList
    End Function



    Public Function GetListByName(ByVal listName As [String]) As SPList
        If Not IsListExisting(listName) Then
            Return Nothing
        End If

        Dim objSPWeb As SPWeb
        objSPWeb = Me.CurrentWeb

        Return objSPWeb.Lists.TryGetList(listName)
    End Function

    Public Function GetListFieldsByListName(ByVal listName As [String]) As SPFieldCollection
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Dim fieldColl As SPFieldCollection = Nothing
        Try
            Dim list As SPList = objSPWeb.Lists.TryGetList(listName)
            If list IsNot Nothing Then
                fieldColl = list.Fields
            End If
        Catch Ex As Exception
            Return Nothing
        End Try
        Return fieldColl
    End Function

    Public Function GetListIDByListName(ByVal listName As [String]) As [String]
        Dim listID As [String] = String.Empty
        Dim objSPList As SPList = Nothing
        Dim objSPWeb As SPWeb
        objSPWeb = Me.CurrentWeb

        objSPList = objSPWeb.Lists.TryGetList(listName)
        If objSPList IsNot Nothing Then
            listID = objSPList.ID.ToString()
        End If
        Return listID
    End Function


    Public Sub GetListItem(ByVal strListName As [String], ByVal strQueryXml As [String], ByRef objSPListItem As SPListItem, ByVal objSPWeb As SPWeb)
        Dim objSPListItemCollection As SPListItemCollection = Nothing

        Try
            objSPListItem = Nothing
            objSPListItemCollection = GetListItems(strListName, strQueryXml, objSPWeb)
            If objSPListItemCollection IsNot Nothing AndAlso objSPListItemCollection.Count > 0 Then
                objSPListItem = objSPListItemCollection(0)
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
    End Sub

    Public Function GetListItemByID(ByVal strListName As [String], ByVal ID As Int32) As SPListItem
        Dim objSPWeb As SPWeb = Me.CurrentWeb.Site.RootWeb
        Return GetListItemByID(strListName, ID, objSPWeb)
    End Function

    Public Function GetListItemByID(ByVal strListName As [String], ByVal ID As Int32, ByVal objSPWeb As SPWeb) As SPListItem
        Dim objSPList As SPList = Nothing
        Dim objSPQuery As SPQuery
        Dim objSPListItemCollection As SPListItemCollection = Nothing
        Dim objSPListItem As SPListItem = Nothing
        Dim strQueryXml As [String] = [String].Empty
        Try
            objSPList = objSPWeb.Lists.TryGetList(strListName)
            If objSPList IsNot Nothing Then
                strQueryXml = "<Where><Eq><FieldRef Name='ID' /><Value Type='Counter'>" & Convert.ToString(ID) & "</Value></Eq></Where>"
                objSPQuery = New SPQuery()
                objSPQuery.ViewAttributes = "Scope='Recursive'"
                objSPQuery.Query = strQueryXml
                objSPListItemCollection = objSPList.GetItems(objSPQuery)
                If objSPListItemCollection.Count > 0 Then
                    objSPListItem = objSPListItemCollection(0)
                End If
            End If
        Catch Ex As Exception
            Throw Ex
        End Try
        Return objSPListItem
    End Function

    Public Function GetListItemByListID(ByVal ListID As [String], ByVal ItemID As Int32, ByVal objSPWeb As SPWeb) As SPListItem
        Dim strListName As [String] = [String].Empty
        strListName = GetListNameByListID(ListID)
        Return GetListItemByID(strListName, ItemID, objSPWeb)
    End Function

    Public Function GetListItemByListID(ByVal ListID As [String], ByVal ItemID As Int32) As SPListItem
        Dim strListName As [String] = [String].Empty
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        strListName = GetListNameByListID(ListID)
        Return GetListItemByID(strListName, ItemID, objSPWeb)
    End Function

    Public Function GetListItems(ByVal strListName As [String], ByVal strQueryXml As [String], ByVal objSPWeb As SPWeb) As SPListItemCollection
        Dim objSPList As SPList = Nothing
        Dim objSPQuery As SPQuery
        Dim objSPListItemCollection As SPListItemCollection = Nothing

        Try

            objSPList = objSPWeb.Lists.TryGetList(strListName)

            If objSPList IsNot Nothing Then
                If Not [String].IsNullOrEmpty(strQueryXml) Then
                    objSPQuery = New SPQuery()
                    objSPQuery.ViewAttributes = "Scope='Recursive'"
                    objSPQuery.Query = strQueryXml
                    objSPListItemCollection = objSPList.GetItems(objSPQuery)
                End If
            End If

        Catch Ex As Exception
        End Try
        Return objSPListItemCollection
    End Function

    Public Function GetListItems(ByVal strListName As [String], ByVal strQueryXml As [String]) As SPListItemCollection
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Return GetListItems(strListName, strQueryXml, objSPWeb)
    End Function


    Public Function GetListNameByListID(ByVal ListID As [String]) As [String]
        Dim strListName As [String] = [String].Empty
        Dim objSPWeb As SPWeb = Me.CurrentWeb
        Dim objSPList As SPList = Nothing
        Try
            objSPList = objSPWeb.Lists.GetList(New Guid(ListID), False, False)
            If objSPList IsNot Nothing Then
                strListName = objSPList.Title
            End If
        Catch Ex As Exception
            Return [String].Empty
        End Try
        Return strListName
    End Function

    Public Function GetListURL(ByVal ListName As [String], Optional ByVal serverRelative As Boolean = False) As [String]
        Dim list As SPList = GetListByName(ListName)
        If list Is Nothing Then
            Return [String].Empty
        End If

        Dim listUrl = list.DefaultViewUrl
        If TypeOf list Is SPDocumentLibrary Then
            Dim idx = listUrl.IndexOf("Forms")
            listUrl = listUrl.Remove(idx)
            If listUrl.EndsWith("/") Then
                listUrl = listUrl.Remove(listUrl.LastIndexOf("/"))
            End If
        Else
            Dim idx = listUrl.LastIndexOf("/")
            listUrl = listUrl.Remove(idx)
        End If
        If Not serverRelative Then
            Dim serverUrl = list.ParentWeb.Url
            serverUrl = serverUrl.Substring(0, serverUrl.IndexOf("/", 7))
            listUrl = serverUrl + listUrl
        End If

        Return listUrl

    End Function

    Public Function GetNewListItemID(ByVal strListName As String, ByVal strFileNameWithoutextension As String) As Integer
        Dim web As SPWeb = Me.CurrentWeb.Site.RootWeb
        Dim objSPList As SPList = Nothing
        Dim objSPListItemCollection As SPListItemCollection = Nothing
        Dim intListItemID As Integer = 0
        Try
            objSPList = web.Lists(strListName)
            objSPListItemCollection = objSPList.Items
            For Each item As SPListItem In objSPListItemCollection
                If item.DisplayName.Equals(strFileNameWithoutextension) Then
                    intListItemID = item.ID
                End If
            Next
        Catch ex As Exception
            Throw ex
        End Try
        Return intListItemID
    End Function

    Public Function GetSPFieldByName(ByVal strListName As [String], ByVal strfieldName As [String]) As SPField
        Dim objSPField As SPField = Nothing
        Dim objSPList As SPList = Nothing
        Try
            objSPList = Me.CurrentWeb.Lists(strListName)
            If objSPList IsNot Nothing Then
                objSPField = objSPList.Fields.GetField(strfieldName)
            End If
        Catch Ex As Exception
            Return Nothing
        End Try
        Return objSPField
    End Function

    Public Function GetSPFieldByName(ByVal fieldColl As SPFieldCollection, ByVal strfieldName As [String]) As SPField
        Dim objSPField As SPField = Nothing

        Try
            If fieldColl IsNot Nothing Then
                objSPField = fieldColl.GetField(strfieldName)
            End If
        Catch Ex As Exception
            Return Nothing
        End Try

        Return objSPField
    End Function

    Public Function IsListExisting(ByVal listName As [String]) As [Boolean]
        Dim listExists As [Boolean] = False
        Dim objSPList As SPList = Nothing
        Dim objSPWeb As SPWeb
        objSPWeb = Me.CurrentWeb

        objSPList = objSPWeb.Lists.TryGetList(listName)
        If objSPList IsNot Nothing Then
            listExists = True
        End If
        Return listExists
    End Function

    Public Function IsListItemExists(ByVal strListName As String, ByVal strFileNameWithoutextension As String) As Boolean
        Dim web As SPWeb = Me.CurrentWeb.Site.RootWeb
        Dim objSPList As SPList = Nothing
        Dim objSPListItemCollection As SPListItemCollection = Nothing
        Dim boolReturn As [Boolean] = False
        Try
            objSPList = web.Lists(strListName)
            objSPListItemCollection = objSPList.Items
            For Each item As SPListItem In objSPListItemCollection
                If item.DisplayName.ToLower().Equals(strFileNameWithoutextension.ToLower()) Then
                    boolReturn = True
                    Exit For
                End If
            Next
        Catch Ex As Exception
            Throw Ex
        End Try
        Return boolReturn
    End Function


    Public Function IsPageExists(ByVal strPath As [String]) As Boolean
        Dim IsExists As [Boolean] = False
        Dim objSPWeb As SPWeb
        Try
            objSPWeb = Me.CurrentWeb
            If objSPWeb.GetFile(strPath).Exists Then
                IsExists = True
            End If

        Catch Ex As Exception
        End Try
        Return IsExists
    End Function



    Public Sub PerformMinorCheckin(ByVal strListName As [String], ByVal ItemID As Int32, ByVal strMessage As [String])
        Dim objListItem As SPListItem = Nothing
        Dim objListOperations As New ListOperations(Me.CurrentWeb)
        Try
            objListItem = objListOperations.GetListItemByID(strListName, ItemID)
            If objListItem.ParentList.EnableVersioning Then
                If objListItem.File.CheckOutType = SPFile.SPCheckOutType.Online Then
                    objListItem.File.CheckIn(strMessage, SPCheckinType.MinorCheckIn)
                End If
            End If
        Catch
        End Try
    End Sub
End Class